docs/development/architecture.org

```diff
@@ -0,0 +1,114 @@
+
+** Loading the Option Tree Data-Structure
+Nix-Gui queries information from the configuration path and =<nixpkgs>= in order to construct a data-structure (=nixui.options.option_tree.OptionTree=) mapping options and attributes to metadata (=nixui.options.option_tree.OptionData=) including their =NixType=, their =OptionDefinition='s:
+- =_type=: A static python-object representation of the options type, informs the frontend of how =FieldWidget= should be generated and input checking should be performed.
+- =system_default_definition=: Provided by =<nixpkgs>=, the default value if nothing is set in the configuration path.
+- =configured_definition=: Provided by the configuration path, overrides =system_default_definition=.
+- =in_memory_definition=: Undefined when the data-structure is loaded, can be mutated and later committed to disk.
+
+=nixui/nix/lib.nix= (invoked by =nixui.options.nix_eval=) contains the function =get_all_nixos_options()= which is used to retrieve metadata to retrieve =OptionData= for the universe of NixOS options.
+
+=nixui/options/parser.py= contains the function =get_all_option_values(module_path)= which
+- retrieves all =attributePath = string representation of nix expression;= pairs in the file by parsing the =nixui.options.syntax_tree.SyntaxTree= constructed using =nix_dump_syntax_tree_json=.
+- resolves the path of all =imports= and recurses
+
+The result of =get_all_option_values= is used to set =OptionData.configured_definition= where applicable.
+
+*** Caveat
+The =OptionTree= has a node for each option in the option tree, as well as attributes that aren't actually options.
+
+For example =users.ldap.loginPam= is a valid nix option. However its "parent", =users.ldap= isn't a valid option. Regardless, =users.ldap= has its own node in the =OptionTree=.
+
+Likewise, =environment.etc."resolv.conf"= isn't an option, it is an attribute of the =environment.etc= option, and it will have a node in the =OptionTree=. This is true of the aforementioned =attribute set of submodules=, as well as =submodule= type options.
+
+=list of= options are part of the option tree despite not having a proper attribute path, e.g. a list of allowed TCP ports in the firewall will have the element =networking.firewall.allowedTCPPorts."[0]"= . This is not the same as setting =networking.firewall.allowedTCPPorts."[0]" = <expression>=, rather, it is the equivalent of =networking.firewall.allowedTCPPorts = [<expression> ...]=.
+
+*** =OptionDefinition=
+Nix expression strings are immediately converted to =nixui.options.option_definition.OptionDefinition= Python objects. The =OptionDefinition= class provides methods to get the python representation of an expression string (e.g. an attribute set is converted to a dict, ="true;"= is converted to =True=).
+
+=OptionDefinition='s also has the =_type= property, which returns the types the current definition is compatible with.
```

Comment:
I don't think you are supposed to use an apostrophe here.

Reply:
Unfortunately org-mode doesn't properly render the inline codeblock properly without an apostrophe, kind of a grammatical hack. I'll look into workarounds in org-mode, otherwise we'll just have to keep it.

Reply:
resolve with https://github.com/nix-gui/nix-gui/issues/226#issuecomment-987532309